File: Java/src/Decode Ways.java
public class Solution
{
  public int numDecodings(String s)
  {
    int length = s.length();
    if (length == 0 || s.charAt(0) == '0')
      return 0;

    int r1 = 1;
    int r2 = 1;
    for (int i = 1; i < length; i++)
    {
      if (s.charAt(i) == '0')
        r1 = 0;
      if (s.charAt(i - 1) == '1' || s.charAt(i - 1) == '2' && s.charAt(i) <= '6')
      {
        r1 = r2 + r1;
        r2 = r1 - r2;
      }
      else
      {
        r2 = r1;
      }
    }

    return r1;
  }
}

File: Python/Decode Ways.py
class Solution:
    # @param s, a string
    # @return an integer
    def numDecodings(self, s):
        length = len(s)
        if length == 0 or s[0] == "0":
            return 0
        
        r1 = 1
        r2 = 1
        
        for i in range(1,length):
            if s[i] == '0': r1 = 0
            if s[i-1] == "1" or s[i-1] == "2" and int(s[i]) <= 6:
                r1 = r2 + r1
                r2 = r1 - r2
            else:
                r2 = r1
        return r1